File: Mechanics_of_Materials-Computer-Problems/chapter-1/p1_c6.m
%%%%%%%%%% P1-C6 %%%%%%%%%%
clc; clear;

f_o_s = input("Factor Of Safety (F.S.): ");
u_s_s = input("Ultimate Shearing Stress [=> MPa]: ");
u_n_s = input("Ultimate Normal Stress [=> MPa]: ");
width_bd = input("Enter the width BD: [=> m]: ");
height_bd = input("Enter the height BD: [=> m]: ");

baseData = struct('label_1', 'AB', 'label_2', 'BC', 'pin_1', 'A', 'pin_2', 'B');
data = cell(3, 1);

for c = 1:3
    if c < 3
        label_name = ['label_' num2str(c)];
        distance = input(['Enter the distance ' baseData.(label_name) ' [=> m]: ']);
    
        pin_name = ['pin_' num2str(c)];
        diameter = input(['Enter the pin diameter ' baseData.(pin_name) ' [=> m]: ']);
    
        dis_name = ['distance_' baseData.(label_name)];
        pin_name = ['pinDiameter_' baseData.(pin_name)];
    
        baseData.(dis_name) = distance;
        baseData.(pin_name) = diameter;
    
        data{c, 1} = struct('diameter', diameter);
    
        if c == 2
            data{1, 1}.p = (baseData.distance_AB / baseData.distance_BC);
            data{2, 1}.p = (baseData.distance_AB / (baseData.distance_AB + baseData.distance_BC));
        end
    else
        data{3, 1}.p = (baseData.distance_AB / (baseData.distance_AB + baseData.distance_BC));
    end
end

for c = 1:3
    if c < 3
        area = (pi / 4) * (data{c, 1}.diameter)^2;
        f = ((2) * (u_s_s) * (area)) / (f_o_s);
    else
        area = width_bd * height_bd;
        f = ((2) * (u_n_s) * (area)) / (f_o_s);
    end

    p =  data{c, 1}.p * f;

    data{c, 1}.area = area;
    data{c, 1}.f = f;
    data{c, 1}.p = p;

    fprintf('\n ------------------------\n   A: %.4e (m²)\n   F: %.4e (N)\n   P: %.3e (N)', area, f, p);

    if c == 3
        [val, min_idx] = min([data{1, 1}.p, data{2, 1}.p, data{3, 1}.p]);
        if min_idx == 3
            mess = "Stress Is Critical In Links";
        elseif min_idx == 1
            mess = "Stress Is Critical In Pin A";
        else
            mess = "Stress Is Critical In Pins B and D";
        end

        fprintf('\n ------------------------\n   P = %.4e , %s.\n', val, mess);
    end
end 

File: Mechanics_of_Materials-Computer-Problems/Code-Mechanics-of-Mat-Semester03/ali/chapter-1/p1_c6.py
import math

fos = float(input("factor of safety (F.S.)= "))
uss = float(input("ultimate shearing stress (MPa)= "))
uns = float(input("ultimate normal stress (MPa)= "))
width_bd = float(input("enter the width BD (m)= "))
height_bd = float(input("enter the height BD (m)= "))

base_data = {'label_1': 'AB', 'label_2': 'BC', 'pin_1': 'A', 'pin_2': 'B'}
data = [{} for _ in range(3)]

for c in range(3):
    if c < 2:
        label_name = f'label_{c+1}'
        distance = float(input(f'enter the distance {base_data[label_name]} (m)= '))
    
        pin_name = f'pin_{c+1}'
        diameter = float(input(f'enter the pin diameter {base_data[pin_name]} (m)= '))
    
        dis_name = f'distance_{base_data[label_name]}'
        pin_name = f'pinDiameter_{base_data[pin_name]}'
    
        base_data[dis_name] = distance
        base_data[pin_name] = diameter
    
        data[c] = {'diameter': diameter}
    
        if c == 1:
            data[0]['p'] = base_data['distance_AB'] / base_data['distance_BC']
            data[1]['p'] = base_data['distance_AB'] / (base_data['distance_AB'] + base_data['distance_BC'])
    else:
        data[2]['p'] = base_data['distance_AB'] / (base_data['distance_AB'] + base_data['distance_BC'])

for c in range(3):
    if c < 2:
        area = (math.pi / 4) * data[c]['diameter']**2
        f = (2 * uss * area) / fos
    else:
        area = width_bd * height_bd
        f = (2 * uns * area) / fos

    p = data[c]['p'] * f

    data[c]['area'] = area
    data[c]['f'] = f
    data[c]['p'] = p

    print(f'\n A = {area:.4e} m²,   F = {f:.4e} N,   P = {p:.3e} N')

    if c == 2:
        val, min_idx = min((data[0]['p'], 0), (data[1]['p'], 1), (data[2]['p'], 2), key=lambda x: x[0])
        if min_idx == 2:
            mess = "stress is critical in links"
        elif min_idx == 0:
            mess = "stress is critical in pin A"
        else:
            mess = "stress is critical in pins B and D"

        print(f'\n P = {val:.4e}, {mess}.\n')